src/diagnostics/gnu-ld.ts

```diff
@@ -4,33 +4,167 @@
 
 import * as vscode from 'vscode';
 
-import { FeedLineResult, oneLess, RawDiagnosticParser } from './util';
+import { oneLess, RawDiagnostic, RawDiagnosticParser, RawRelated, FeedLineResult } from './util';
 
-export const REGEX = /^(.*):(\d+)\s?:\s+(.*[^\]])$/;
+enum MatchType {
+    Full,
+    File,
+    Line,
+    Column,
+    Severity,
+    Message
+}
+
+const regexPatterns: RegexPattern[] = [
+    {   // path/to/ld[.exe]:[ ]path/to/file:line: severity: message
+        regexPattern: /^(?:.*ld(?:\.exe)?:)(?:\s*)?(.+):(\d+):\s+(?:fatal )?(\w+):\s+(.+)/,
+        matchTypes: [MatchType.Full, MatchType.File, MatchType.Line, MatchType.Severity, MatchType.Message]
+    },
+    {   // path/to/ld[.exe]:[ ]path/to/file:line: message
+        regexPattern: /^(?:.*ld(?:\.exe)?\:)(?:\s*)?(.+):(\d+):\s+(.+)/,
+        matchTypes: [MatchType.Full, MatchType.File, MatchType.Line, MatchType.Message]
+    },
+    {   // path/to/ld[.exe]: severity: message
+        regexPattern: /^(.*ld(?:\.exe)?):\s+(?:fatal )?(\w+):\s+(.+)/,
+        matchTypes: [MatchType.Full, MatchType.File, MatchType.Severity, MatchType.Message]
+    },
+    {   // path/to/ld[.exe]: message (without trailing colon)
+        regexPattern: /^(.*ld(?:\.exe)?):\s+(.+)(?<!:)$/,
+        matchTypes: [MatchType.Full, MatchType.File, MatchType.Message]
+    },
+    {   // /path/to/file:line: message (without "[fatal] severity:" or trailing colon)
+        regexPattern: /^(.+?):(\d+):\s+(?!fatal\s+\w+:)(?!\w+:)(.+)(?<!:)$/,
+        matchTypes: [MatchType.Full, MatchType.File, MatchType.Line, MatchType.Message]
+    }
+];
+
+interface RegexPattern {
+    regexPattern: RegExp;
+    matchTypes: MatchType[];
+}
+
+interface PendingTemplateBacktrace {
+    rootInstantiation: string;
+    requiredFrom: RawRelated[];
+}
 
 export class Parser extends RawDiagnosticParser {
+    private _prevDiag?: RawDiagnostic;
+
+    private _pendingTemplateError?: PendingTemplateBacktrace;
+
     doHandleLine(line: string) {
-        // Try to parse for GNU ld
-        if (line.startsWith('make')) {
-            // This is a Make error. It may *look* like an LD error, so we abort early
-            return FeedLineResult.NotMine;
+        let mat = /(.*): (In instantiation of.+)/.exec(line);
```

Comment:
Could you provide some helpful comments that explain why we are doing some of these `.exec`s?

Reply:
I replaced the gnu-ld.ts code with gcc.ts, assuming the template and backtrace code is for parsing non-C-language related stuff. Your comment made me look into it, and it turns out they're not needed for linker errors.

The code is for C++ template errors, which cause multiline error messages, such as:
```
main.cpp: In instantiation of ‘void fkt(Foo) [with Foo = int]’:
main.cpp:5:7:   required from here
main.cpp:2:6: warning: unused parameter ‘f’ [-Wunused-parameter]
````
So 
- On the first call, `let mat = /(.*): (In instantiation of.+)/.exec(line);` detects the startline, creates a this._pendingTemplateError and returns early
- On the second call, `this._pendingTemplateError` is `!= NULL`, so the `mat = /(.*):(\d+):(\d+):(  +required from.+)/.exec(line);` is used to detect the second line
- On the third call, the actual diagnostic line is parsed and the template error is completed

I removed the code from the linker diagnostics.

Reply:
@gcampbell-msft Am I expected to "Resolve conversation" or do you do that when you are satisfied?

Reply:
@0xemgy Could you include some of these comments you made above in the source code?

Reply:
I added comments to both the gcc and ld parser files.